File: Frontend/src/services/complaintService.js
import { apiClient } from "./authService";

/**
 * Complaint Service
 * Dịch vụ quản lý complaints/khiếu nại
 */
const complaintService = {
  /**
   * Seller xem tất cả complaints về mình
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getComplaintsAboutMe() {
    try {
      const response = await apiClient.get("/api/complaints/complaints-about-me");
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: Array.isArray(data) ? data : [],
        message: response?.data?.message || "Lấy danh sách complaints thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        data: [],
        message: `Lỗi tải complaints (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },

  /**
   * Buyer xem tất cả complaints của mình
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getMyComplaints() {
    try {
      const response = await apiClient.get("/api/complaints/my-complaints");
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: Array.isArray(data) ? data : [],
        message: response?.data?.message || "Lấy danh sách complaints thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        data: [],
        message: `Lỗi tải complaints (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },

  /**
   * Lấy chi tiết một complaint
   * @param {number} complaintId - ID của complaint
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getComplaintById(complaintId) {
    try {
      const response = await apiClient.get(`/api/complaints/${complaintId}`);
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Lấy chi tiết complaint thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        message: `Lỗi tải complaint (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },

  /**
   * Admin/Staff lấy tất cả complaints
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getAllComplaints() {
    try {
      const response = await apiClient.get("/api/complaints/admin/all");
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: Array.isArray(data) ? data : [],
        message: response?.data?.message || "Lấy tất cả complaints thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        data: [],
        message: `Lỗi tải complaints (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },

  /**
   * Buyer tạo complaint mới
   * @param {Object} complaintData - Dữ liệu complaint {contractId, title, description, category, evidenceImages}
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async createComplaint(complaintData) {
    try {
      const formData = new FormData();
      formData.append("contractId", complaintData.contractId);
      formData.append("title", complaintData.title);
      formData.append("description", complaintData.description);
      formData.append("category", complaintData.category);
      
      // Append evidence images nếu có
      if (complaintData.evidenceImages && complaintData.evidenceImages.length > 0) {
        complaintData.evidenceImages.forEach((file, index) => {
          formData.append("evidenceImages", file);
        });
      }

      const response = await apiClient.post("/api/complaints", formData, {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      });
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Tạo complaint thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        message: `Lỗi tạo complaint (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },

  /**
   * Kiểm tra buyer và seller đã có giao dịch hoàn thành
   * @param {number} sellerId - ID của seller
   * @returns {Promise<{success: boolean, data?: boolean, message?: string}>}
   */
  async checkCompletedTransaction(sellerId) {
    try {
      const response = await apiClient.get(`/api/complaints/check-transaction/${sellerId}`);
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Kiểm tra giao dịch thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        message: `Lỗi kiểm tra giao dịch (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },

  /**
   * Admin/Staff bắt đầu xem xét complaint
   * @param {number} complaintId - ID của complaint
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async startReview(complaintId) {
    try {
      const response = await apiClient.put(`/api/complaints/admin/${complaintId}/start-review`);
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Bắt đầu xem xét thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        message: `Lỗi bắt đầu xem xét (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },

  /**
   * Admin/Staff giải quyết complaint
   * @param {number} complaintId - ID của complaint
   * @param {Object} resolutionData - Dữ liệu giải quyết {status, staffNotes}
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async resolveComplaint(complaintId, resolutionData) {
    try {
      const response = await apiClient.put(`/api/complaints/admin/${complaintId}/resolve`, resolutionData);
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Giải quyết complaint thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      return {
        success: false,
        message: `Lỗi giải quyết complaint (${status || "network"}): ${
          backendMessage || "Không rõ"
        }`,
      };
    }
  },
};

export default complaintService;

File: Backend/src/main/java/com/example/backend/service/ComplaintService.java
package com.example.backend.service;

import com.example.backend.dto.request.AdminResolveComplaintRequest;
import com.example.backend.dto.request.ComplaintRequest;
import com.example.backend.dto.response.ComplaintResponse;
import com.example.backend.entity.Complaint;
import com.example.backend.entity.Contract;
import com.example.backend.entity.User;
import com.example.backend.enums.ComplaintStatus;
import com.example.backend.enums.ContractStatus;
import com.example.backend.exception.AppException;
import com.example.backend.exception.ErrorCode;
import com.example.backend.mapper.ComplaintMapper;
import com.example.backend.repository.ComplaintRepository;
import com.example.backend.repository.ContractRepository;
import com.example.backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class ComplaintService {
    
    private final ComplaintRepository complaintRepository;
    private final ContractRepository contractRepository;
    private final UserRepository userRepository;
    private final ComplaintMapper complaintMapper;
    private final CloudinaryService cloudinaryService;
    
    /**
     * Tạo complaint mới - chỉ cho phép khi đã có giao dịch hoàn thành
     */
    @Transactional
    public ComplaintResponse createComplaint(ComplaintRequest request, Long buyerId) {
        // Lấy contract
        Contract contract = contractRepository.findById(request.getContractId())
                .orElseThrow(() -> new AppException(ErrorCode.CONTRACT_NOT_FOUND));
        
        // Lấy buyer
        User buyer = userRepository.findById(buyerId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        // Kiểm tra buyer có phải là buyer trong contract không
        if (contract.getBuyer().getId() != buyerId) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }
        
        // Kiểm tra contract đã hoàn thành chưa (status = COMPLETED và deliveryCompleted = true)
        if (contract.getStatus() == ContractStatus.COMPLETED || !contract.getDeliveryCompleted()) {
            throw new AppException(ErrorCode.CONTRACT_NOT_COMPLETED);
        }
        
        // Kiểm tra đã có complaint cho contract này chưa
        if (complaintRepository.existsByContract(contract)) {
            throw new AppException(ErrorCode.COMPLAINT_ALREADY_EXISTS);
        }
        
        // Upload evidence images nếu có
        List<String> evidenceUrls = null;
        if (request.getEvidenceImages() != null && !request.getEvidenceImages().isEmpty()) {
            evidenceUrls = new ArrayList<>();
            for (MultipartFile file : request.getEvidenceImages()) {
                String url = cloudinaryService.upload(file);
                evidenceUrls.add(url);
            }
        }
        
        // Tạo complaint
        Complaint complaint = Complaint.builder()
                .buyer(buyer)
                .seller(contract.getSeller())
                .product(contract.getProduct())
                .contract(contract)
                .title(request.getTitle())
                .description(request.getDescription())
                .category(request.getCategory())
                .evidenceUrls(evidenceUrls)
                .build();
        
        Complaint savedComplaint = complaintRepository.save(complaint);
        
        log.info("Complaint created successfully: {}", savedComplaint.getId());
        return complaintMapper.toComplaintResponse(savedComplaint);
    }
    
    /**
     * Lấy tất cả complaint của một buyer
     */
    public List<ComplaintResponse> getComplaintsByBuyer(Long buyerId) {
        User buyer = userRepository.findById(buyerId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        List<Complaint> complaints = complaintRepository.findByBuyer(buyer);
        return complaintMapper.toComplaintResponseList(complaints);
    }
    
    /**
     * Lấy tất cả complaint của một seller
     */
    public List<ComplaintResponse> getComplaintsBySeller(Long sellerId) {
        User seller = userRepository.findById(sellerId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        List<Complaint> complaints = complaintRepository.findBySeller(seller);
        return complaintMapper.toComplaintResponseList(complaints);
    }
    
    /**
     * Lấy chi tiết một complaint
     */
    public ComplaintResponse getComplaintById(Long complaintId) {
        Complaint complaint = complaintRepository.findById(complaintId)
                .orElseThrow(() -> new AppException(ErrorCode.COMPLAINT_NOT_FOUND));
        
        return complaintMapper.toComplaintResponse(complaint);
    }
    
    /**
     * Lấy tất cả complaint (cho admin)
     */
    public List<ComplaintResponse> getAllComplaints() {
        List<Complaint> complaints = complaintRepository.findAll();
        return complaintMapper.toComplaintResponseList(complaints);
    }
    
    /**
     * Kiểm tra buyer và seller đã có giao dịch hoàn thành chưa
     */
    public boolean hasCompletedTransaction(Long buyerId, Long sellerId) {
        User buyer = userRepository.findById(buyerId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        User seller = userRepository.findById(sellerId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        return complaintRepository.hasCompletedTransaction(buyer, seller);
    }
    
    /**
     * Admin giải quyết complaint
     */
    @Transactional
    public ComplaintResponse adminResolveComplaint(Long complaintId, AdminResolveComplaintRequest request) {
        // Lấy complaint
        Complaint complaint = complaintRepository.findById(complaintId)
                .orElseThrow(() -> new AppException(ErrorCode.COMPLAINT_NOT_FOUND));
        
        // Kiểm tra complaint chưa được giải quyết
        if (complaint.getStatus() != ComplaintStatus.PENDING && complaint.getStatus() != ComplaintStatus.UNDER_REVIEW) {
            throw new AppException(ErrorCode.COMPLAINT_ALREADY_RESOLVED);
        }
        
        // Cập nhật status và staff notes
        complaint.setStatus(request.getStatus());
        complaint.setStaffNotes(request.getStaffNotes());
        
        // Nếu đã giải quyết thì set resolvedAt
        if (request.getStatus() == ComplaintStatus.RESOLVED_BUYER_FAVOR || 
            request.getStatus() == ComplaintStatus.RESOLVED_SELLER_FAVOR ||
            request.getStatus() == ComplaintStatus.CLOSED) {
            complaint.setResolvedAt(LocalDateTime.now());
        }
        
        Complaint savedComplaint = complaintRepository.save(complaint);
        
        log.info("Complaint {} resolved by admin with status: {}", complaintId, request.getStatus());
        return complaintMapper.toComplaintResponse(savedComplaint);
    }
    
    /**
     * Admin chuyển complaint sang trạng thái đang xem xét
     */
    @Transactional
    public ComplaintResponse adminStartReview(Long complaintId) {
        Complaint complaint = complaintRepository.findById(complaintId)
                .orElseThrow(() -> new AppException(ErrorCode.COMPLAINT_NOT_FOUND));
        
        if (complaint.getStatus() != ComplaintStatus.PENDING) {
            throw new AppException(ErrorCode.COMPLAINT_NOT_PENDING);
        }
        
        complaint.setStatus(ComplaintStatus.UNDER_REVIEW);
        Complaint savedComplaint = complaintRepository.save(complaint);
        
        log.info("Complaint {} moved to UNDER_REVIEW by admin", complaintId);
        return complaintMapper.toComplaintResponse(savedComplaint);
    }
}
